content/docs/learn/typed-errors/working-with-typed-errors.md

```diff
@@ -661,90 +808,21 @@ which follow a short-circuiting approach.
-Within that class, a `bind` function can be defined to short-circuit any encountered `Failure` or `Loading` case or otherwise return the `Content` value.
-
-```kotlin
-@JvmInline
-value class LceRaise<E>(val raise: Raise<Lce<E, Nothing>>) : Raise<Lce<E, Nothing>> by raise {
-  fun <A> Lce<E, A>.bind(): A =  when (this) {
-    is Lce.Content -> value
-    is Lce.Failure -> raise.raise(this)
-    Lce.Loading -> raise.raise(Lce.Loading)
-  }
-}
-```
-
-All that is required now is a DSL function. We can use the `recover` or `fold` function to summon an instance of `RaiseLce<E, Nothing>` from the `Raise` type class.
-We wrap the `block` in an `Lce.Content` value and return any encountered `Lce<E, Nothing>` value. We can call `block` by wrapping `Raise<Lce<E, Nothing>>` in `LceRaise`.
-
-```kotlin
-@OptIn(ExperimentalTypeInference::class)
-inline fun <E, A> lce(@BuilderInference block: LceRaise<E>.() -> A): Lce<E, A> =
-  recover({ Lce.Content(block(LceRaise(this))) }) { e: Lce<E, Nothing> -> e }
-```
+## Summary
 
-We can now use this DSL to compose our computations and `Lce` values in the same way as we've discussed above in this document.
-Furthermore, since this DSL is built on top of `Raise`, we can use all the functions we've discussed above.
+At this point we can summarize the advantages that typed errors offer over using exceptions:
 
-```kotlin
-fun example() {
-  lce {
-    val a = Lce.Content(1).bind()
-    val b = Lce.Content(1).bind()
-    a + b
-  } shouldBe Lce.Content(2)
-
-  lce {
-    val a = Lce.Content(1).bind()
-    ensure(a > 1) { Lce.Failure("a is not greater than 1") }
-    a + 1
-  } shouldBe Lce.Failure("a is not greater than 1")
-}
-```
-<!--- KNIT example-typed-errors-16.kt -->
-<!--- TEST assert -->
+- **Type Safety:** Typed errors allow the compiler to find type mismatches early, making it easier to catch bugs before they make it to production. However, with exceptions, the type information is lost, making it more difficult to detect errors at compile-time.
 
-If we'd used _context receivers_, defining this DSL would be even more straightforward, and we could use the `Raise` type class directly.
+- **Predictability:** When using typed errors, the possible error conditions are explicitly listed in the type signature of a function. This makes it easier to understand the possible error conditions and write tests covering all error scenarios.
 
-```kotlin
-context(Raise<Lce<E, Nothing>>)
-fun <E, A> Lce<E, A>.bind(): A =  when (this) {
-  is Lce.Content -> value
-  is Lce.Failure -> raise(this)
-  Lce.Loading -> raise(Lce.Loading)
-}
+- **Composability:** Typed errors can be easily combined and propagated through a series of function calls, making writing modular, composable code easier. With exceptions, ensuring errors are correctly propagated through a complex codebase can be difficult. Patterns like accumulation, which are at your fingertips using typed errors, become quite convoluted using exceptions.
```

Comment:
Another suggestion to the readme @serras. I've got some feedback from colleagues that when they reads this documentation, they were looking to find a code example that illustrate computation chaining for a production use case. I remember something like that used to exist in the old docs about comprehension.

Having said that, may I suggest adding a snippet along these lines. This is the same code I put in kotlinlang https://kotlinlang.slack.com/archives/C0B9K7EP2/p1680694416774339?thread_ts=1680388003.721149&cid=C0B9K7EP2. Feel free to modify this as appropriate.. Notice that I've also used sealed class as the failure case. I noticed most (i guess > 80%) production usecases that we've encountered uses sealed classes as well to model the failures. I'm not going to be surprised if many users out there also do the same. I hope that can give something more concrete for users.

```kotlin
context(Raise<UpdateUserFailure>) // this is a different error boundary
suspend fun updateUser(...): User = ...

context(Raise<ProvisioningFailure>)
suspend fun checkManagedStatus(...): Unit = ensure(user.isManaged) { 
  ProvisioningFailure.NotManaged 
}

context(Raise<ProvisioningFailure>)
suspend fun checkProvisioningPolicy(...): Unit = ...

context(Raise<ProvisioningFailure>)
suspend fun provision(): User {
  checkManagedStatus(...)
  checkProvisioningPolicy(...)

  val updatedUser = recover({ updateUser(...) }) {
    when (it) {
      is UserNotFound -> raise(ProvisioningFailure.InvalidUser)
      is UpdateRejected -> raise(ProvisioningFailure.InvalidUpdate)
    }
  }

  updatedUser
}
```

We can also encourage user to prepare for context receiver accordingly by providing a version of the composition in `either { }`. 
```kotlin
suspend fun provision(): Either<ProvisioningFailure, User> = either {
  checkManagedStatus(...).bind()
  checkProvisioningPolicy(...).bind()

  val updatedUser = recover({ updateUser(...).bind() }) {
    when (it) {
      is UserNotFound -> raise(ProvisioningFailure.InvalidUser)
      is UpdateRejected -> raise(ProvisioningFailure.InvalidUpdate)
    }
  }

  updatedUser
}
```

Reply:
Thanks for the very concrete example, @myuwono. We got similar feedback from other people, which only reinforces the fact that our docs are lacking in that respect. I made some changes in this document to clarify the situation, do you think the read better now?

I'm hesitant, though, to make the introduction even longer. Maybe we should have a full worked-out example where we describe these ideas, like how to design your error hierarchy, the different boundaries, and so on?

Reply:
That's a brilliant idea.